moveit_ros/moveit_servo/src/servo_calcs.cpp

```diff
@@ -787,24 +787,25 @@ bool ServoCalcs::enforcePositionLimits()
   {
     // Halt if we're past a joint margin and joint velocity is moving even farther past
     double joint_angle = 0;
-    for (std::size_t c = 0; c < original_joint_state_.name.size(); ++c)
+    std::size_t idx = 0;
+    for (idx = 0; idx < original_joint_state_.name.size(); ++idx)
     {
-      if (original_joint_state_.name[c] == joint->getName())
+      if (original_joint_state_.name[idx] == joint->getName())
       {
-        joint_angle = original_joint_state_.position.at(c);
+        joint_angle = original_joint_state_.position.at(idx);
         break;
       }
     }
     if (!current_state_->satisfiesPositionBounds(joint, -parameters_.joint_limit_margin))
     {
-      const std::vector<moveit_msgs::JointLimits> limits = joint->getVariableBoundsMsg();
+      const std::vector<moveit_msgs::JointLimits>& limits = joint->getVariableBoundsMsg();
 
       // Joint limits are not defined for some joints. Skip them.
       if (!limits.empty())
       {
-        if ((current_state_->getJointVelocities(joint)[0] < 0 &&
+        if ((internal_joint_state_.velocity[idx] < 0 &&
```

Comment:
If I understand right, this PR uses `internal_joint_state_` because it takes the low-pass filtering of the signal into account. `current_state_` that was used previously comes from the planning scene monitor so it isn't aware of that filtering.

I'm gonna suggest a comment here then LGTM. I'll try to get one other person to review it before merging.

Reply:
```suggestion
        // Use `internal_joint_state_` rather than `current_state_` so that low-pass filtering is accounted for
        if ((internal_joint_state_.velocity[idx] < 0 &&
```

Reply:
So I'm actually using the `internal_joint_state_` because that contains the velocities that we want to command (it's passed in to the `composeJointTrajMessage` function) and will fix that issue described in the description. Meaning, we look at the velocities we want to command to see if they will put the joint even further from the joint limit or not. Looking at the current velocities is problematic since it doesn't give us that insight which leads to the issue described in the description where if the velocities were already zero due to a "halt" from the previous iteration, positions were not enforced since a zero-velocity case was considered OK thereby causing the next command to be nonzero.